src/client/mod.rs

```diff
@@ -0,0 +1,592 @@
+                --- ...
+            
+            After writing, every metric is given ownership
+            of the respective memory-mapped slice that contains
+            the metric's value. This is to ensure that the metric
+            is *only* able to write to it's value slice when updating
+            it's value.
+        */
+ 
+        {
+            let mut cur = Cursor::new(unsafe { mmap_view.as_mut_slice() });
+            self.write_mmv_header(&mut cur)?;
+            self.write_metric_toc_block(&mut cur)?;
+            self.write_values_toc_block(&mut cur)?;
+            self.write_strings_toc_block(&mut cur)?;
+        }
+
+        self.wi.mmap_view = Some(mmap_view);
+        Ok(self)
+    }
+
+    fn write_mmv_header(&mut self, c: &mut Cursor<&mut [u8]>) -> io::Result<()> {    
+        // MMV\0
+        c.write_all(CString::new("MMV")?.to_bytes_with_nul())?;
+        // version
+        c.write_u32::<Endian>(1)?;
+        // generation1
+        self.wi.gen = time::now().to_timespec().sec;
+        c.write_i64::<Endian>(self.wi.gen)?;
+        // generation2
+        self.wi.gen2off = c.position();
+        c.write_i64::<Endian>(0)?;
+        // no. of toc blocks
+        c.write_u32::<Endian>(TOC_BLOCK_COUNT as u32)?;
+        // flags
+        c.write_u32::<Endian>(self.flags.bits())?;
+        // pid
+        c.write_i32::<Endian>(get_process_id())?;
+        // cluster id
+        c.write_u32::<Endian>(self.cluster_id)
+    }
+    
+    fn write_metric_toc_block(&mut self, c: &mut Cursor<&mut [u8]>) -> io::Result<()> {
+        // section type
+        c.write_u32::<Endian>(3)?;
+        // no. of entries
+        c.write_u32::<Endian>(self.wi.n_metrics as u32)?;
+        // section offset
+        self.wi.metric_sec_off = HDR_LEN + TOC_BLOCK_LEN*TOC_BLOCK_COUNT;
+        c.write_u64::<Endian>(self.wi.metric_sec_off)
+    }
+
+    fn write_values_toc_block(&mut self, c: &mut Cursor<&mut [u8]>) -> io::Result<()> {
+        // section type
+        c.write_u32::<Endian>(4)?;
+        // no. of entries
+        c.write_u32::<Endian>(self.wi.n_metrics as u32)?;
+        // section offset
+        self.wi.value_sec_off = self.wi.metric_sec_off + METRIC_BLOCK_LEN*self.wi.n_metrics;
+        c.write_u64::<Endian>(self.wi.value_sec_off)
+    }
+
+    fn write_strings_toc_block(&mut self, c: &mut Cursor<&mut [u8]>) -> io::Result<()> {
+        // section type
+        c.write_u32::<Endian>(5)?;
+        // no. of entries
+        self.wi.n_strings_off = c.position();
+        c.write_u32::<Endian>(self.wi.n_strings as u32)?;
+        // section offset
+        self.wi.string_sec_off = self.wi.value_sec_off + VALUE_BLOCK_LEN*self.wi.n_metrics;
+        c.write_u64::<Endian>(self.wi.string_sec_off)
+    }
+
+    pub fn register_metric<T: MetricType + Clone>(&mut self, m: &mut Metric<T>) -> io::Result<&mut Client> {
+
+        // TODO: return custom error instead of panicing
+        assert!(self.wi.metric_idx < self.wi.n_metrics);
```

Comment:
Yes, return a custom error which has an `io::Error` as a variant and also implements `From<io::Error>`